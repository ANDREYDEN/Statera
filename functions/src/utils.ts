import * as vision from '@google-cloud/vision'
import { VerticalSegment } from './types/geometry'
import _ from 'lodash'

export function verticalSegment(
    label: vision.protos.google.cloud.vision.v1.IEntityAnnotation
): VerticalSegment {
  const corners = label.boundingPoly?.vertices ?? []
  const sortedCorners = corners.sort(
      (corner, otherCorner) => (corner?.y ?? 0) - (otherCorner?.y ?? 0)
  )
  const getY = (i: number) => sortedCorners[i]?.y ?? 0
  const getX = (i: number) => sortedCorners[i]?.x ?? 0
  return {
    yTop: (getY(0) + getY(1)) / 2,
    yBottom: (getY(2) + getY(3)) / 2,
    x: (getX(0) + getX(1) + getX(2) + getX(3)) / 4,
  }
}

export function stripSku(sku: string): string {
  const matches = sku.match(/0*(.*?)(R|$)/)
  if (!matches) return sku
  return matches[1]
}

export function toPascalCase(str: string): string {
  if (!str.match(/\w+/)) return str

  return _.camelCase(str).replace(/^(.)/, _.toUpper)
}

export function propertyChanged<T>(obj1: T, obj2: T, propertyName: keyof T): boolean {
  return JSON.stringify(obj1[propertyName]) != JSON.stringify(obj2[propertyName])
}
