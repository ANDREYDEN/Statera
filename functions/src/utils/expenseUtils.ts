import { Expense } from '../types/expense'

export function calculateStage(expense: Expense, assigneeId: string) {
  if (expense.finalizedDate) return 2

  if (expense.unmarkedAssigneeIds.includes(assigneeId)) return 0

  return 1
}
