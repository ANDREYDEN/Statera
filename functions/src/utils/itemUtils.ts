import { Item } from '../types/item'

export function isItemCompleted(item: Item) : boolean {
  if (item.partition > 1) {
    const confirmedParts = item.assignees
      .map((a) => a.parts)
      .reduce((acc, cur) => (acc ?? 0) + (cur ?? 0), 0)
    return confirmedParts === item.partition
  }

  return item.assignees.every((a) => a.parts != null)
}
