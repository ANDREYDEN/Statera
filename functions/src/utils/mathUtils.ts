export const avg = (...params: number[]) => params.reduce((acc, p) => acc + p, 0) / params.length

export const sqr = (x: number) => Math.pow(x, 2)
