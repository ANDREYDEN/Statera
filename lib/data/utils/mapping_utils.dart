Map<String, Map<String, double>>? mapBalance(dynamic balanceData) {
  if (balanceData == null) return null;

  return Map<String, Map<String, double>>.from(
    balanceData.map(
      (uid, balance) => MapEntry(
        uid,
        Map<String, double>.from(
          (balance as Map<String, dynamic>).map((otherUid, value) =>
              MapEntry(otherUid, double.tryParse(value.toString()))),
        ),
      ),
    ),
  );
}
