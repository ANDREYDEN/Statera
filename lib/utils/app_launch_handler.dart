import 'dart:developer';

import 'package:firebase_messaging/firebase_messaging.dart';
import 'package:flutter/cupertino.dart';
import 'package:go_router/go_router.dart';

class AppLaunchHandler {
  static const Duration cooldown = Duration(seconds: 1);

  static DateTime? _lastLaunchHandledAt;

  static bool get canHandleLaunch =>
      _lastLaunchHandledAt == null ||
      DateTime.now().difference(_lastLaunchHandledAt!) > cooldown;

  static void _ensureCanHandleLaunch() {
    if (!canHandleLaunch) return;
    _lastLaunchHandledAt = DateTime.now();
  }

  /// Handles the notification [message] (tapping on a notification) from a particular app [context].
  /// This method ensures that each notification will be handled exactly once
  /// (ignoring all invocations for a given period of time)
  static void handleNotificationMessage(
    RemoteMessage message,
    BuildContext context,
  ) {
    _ensureCanHandleLaunch();
    log('handling message ${message.data}');

    final path = getPath(message);
    if (path == null) return;

    context.go(path);
  }

  /// Handles the dynamic link [path] from a particular app [context].
  /// This method ensures that each dynamic link will be handled exactly once
  /// (ignoring all invocations for a given period of time)
  static void handleDynamicLink(String path, BuildContext context) {
    _ensureCanHandleLaunch();
    log('handling dynamic link $path');

    context.go(path);
  }

  static String? getPath(RemoteMessage? message) {
    if (message == null) return null;
    switch (message.data['type']) {
      case 'expense_created':
      case 'expense_finalized':
        if (message.data['expenseId'] != null) {
          return '/expense/${message.data['expenseId']}';
        }
        break;
      case 'group_debt_threshold_reached':
      case 'expense_completed':
        if (message.data['groupId'] != null) {
          return '/group/${message.data['groupId']}';
        }
        break;
      default:
        return null;
    }

    return null;
  }
}
