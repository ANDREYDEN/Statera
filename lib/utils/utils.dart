export 'constants.dart';
export 'formatters.dart';
export 'helpers.dart';
export 'theme.dart';
export 'app_launch_handler.dart';