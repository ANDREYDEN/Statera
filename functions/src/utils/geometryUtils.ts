
import { google } from '@google-cloud/vision/build/protos/protos'
import { BoxWithText, Vector } from '../types/geometry'
import { sqr } from './mathUtils'

export type IEntityAnnotation = google.cloud.vision.v1.IEntityAnnotation
type IVertex = google.cloud.vision.v1.IVertex

export function toBoxWithText(annotation: IEntityAnnotation): BoxWithText {
  const corners = annotation.boundingPoly?.vertices ?? []
  const [topLeft, topRight, bottomRight, bottomLeft] = corners

  const y = (v: IVertex) => v?.y ?? 0
  const x = (v: IVertex) => v?.x ?? 0
  const mid = (a: number, b: number) => (a + b) / 2

  const top: Vector = {
    x: mid(x(topLeft), x(topRight)),
    y: mid(y(topLeft), y(topRight)),
  }
  const bottom: Vector = {
    x: mid(x(bottomLeft), x(bottomRight)),
    y: mid(y(bottomLeft), y(bottomRight)),
  }
  const left: Vector = {
    x: mid(x(bottomLeft), x(topLeft)),
    y: mid(y(bottomLeft), y(topLeft)),
  }
  const right: Vector = {
    x: mid(x(bottomRight), x(topRight)),
    y: mid(y(bottomRight), y(topRight)),
  }
  const center: Vector = {
    x: mid(top.x, bottom.x),
    y: mid(left.y, right.y),
  }
  return {
    top, bottom, right, left, center,
    content: annotation.description,
  }
}

export function distanceToRow(row: BoxWithText[], box2: BoxWithText): number {
  const line = [row[0].left, row[row.length - 1].right]
  const normalizedLine = sub(line[1], line[0])
  const normalizedCenter = sub(box2.center, line[0])

  return distanceToVectorLine(normalizedCenter, normalizedLine)
}

export function add(a: Vector, b: Vector): Vector {
  return { x: a.x + b.x, y: a.y + b.y }
}

export function sub(a: Vector, b: Vector): Vector {
  return { x: a.x - b.x, y: a.y - b.y }
}

export function len(v: Vector): number {
  return Math.sqrt(sqr(v.x) + sqr(v.y))
}

export function tan(v: Vector): number {
  return v.y / v.x
}

export function height(box: BoxWithText): number {
  return len(sub(box.top, box.bottom))
}

function distanceToVectorLine(point: Vector, vector: Vector): number {
  const d1 = len(vector)
  return Math.abs(point.x*vector.y - vector.x*point.y) / d1
}
