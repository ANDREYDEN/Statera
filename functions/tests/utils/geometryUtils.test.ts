import { IEntityAnnotation, toBoxWithText } from '../../src/utils/geometryUtils'

describe('geometryUtils', () => {
  describe('toBoxWithText', () => {
    it('converts annotation to box with text', () => {
      const annotation: IEntityAnnotation = {
        boundingPoly: {
          vertices: [
            {
              x: 103,
              y: 288,
            },
            {
              x: 145,
              y: 276,
            },
            {
              x: 150,
              y: 293,
            },
            {
              x: 108,
              y: 305,
            },
          ],
        },
        description: 'E & OE',
      }
      const box = toBoxWithText(annotation)

      expect(box).toMatchSnapshot()
    })
  })
})
