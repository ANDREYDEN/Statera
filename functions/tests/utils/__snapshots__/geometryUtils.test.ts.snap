// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geometryUtils toBoxWithText converts annotation to box with text 1`] = `
{
  "bottom": {
    "x": 129,
    "y": 299,
  },
  "center": {
    "x": 126.5,
    "y": 290.5,
  },
  "content": "E & OE",
  "left": {
    "x": 105.5,
    "y": 296.5,
  },
  "right": {
    "x": 147.5,
    "y": 284.5,
  },
  "top": {
    "x": 124,
    "y": 282,
  },
}
`;
